### DISCUSSION TITLE
Issue with Predefined Data Types in .idl File Using Fast DDS

### DISCUSSION THREAD

**Discussion by Akul1903** [eProsima/Fast-RTPS-docs#900] [General]
Dear Fast DDS Team,

I am encountering an issue while writing code in the .idl file. Only two types of predefined data are working, while the rest are showing as “not defined.”

Here is an example of the .idl file that is working:

```
struct helloworld
{
    unsigned long index;
    string message;
};
```

However, other data types are not working. I am compiling with the fastddsgen path along with the -example CMake file.idl command.

Could you please assist me in resolving this issue?

Thank you for your support.

Regards
Arpit Kulshrestha

**Comment by JesusPoderoso** [comment]
Hi @Akul1903, thanks for using Fast DDS.

This [documentation section](https://fast-dds.docs.eprosima.com/en/latest/fastddsgen/dataTypes/dataTypes.html#supported-idl-types) describes the supported types that can be defined in *idl* files.

What are the data types that are not working for you?

**Reply by Akul1903** [reply]
Hi @JesusPoderoso thanks for your words

I am working on these two  data types:
unsigned long,string


And One more doubt By any chance fastdds is a license product or a free open source product

**Reply by Akul1903** [reply]
Hi @JesusPoderoso 

Kindly help its urgent

while working on fast dds publisher subscriber model , I get stuck in a problem
I am working on one publisher and 2 subscriber model in which each subscriber has different data members.

For example:

Publisher idl file:
enum Status {
    ACTIVE,
    INACTIVE,
    PENDING
};

struct user {
    char id;
    string name;
    string email;
};

struct helloworld {
    unsigned long index;
    string message;
    Status ss; 
    user userInfo; 
};


Subscriber 1 idl file:
struct user {
    char id;
    string name;
    string email;
};

struct helloworld {
    unsigned long index;
    string message;
    user userInfo; 
};


and subscriber 2 idl file:
enum Status {
    ACTIVE,
    INACTIVE,
    PENDING
};


struct helloworld {
    unsigned long index;
    string message;
    Status ss; 
};


SO BASICALLY THIS IS MY QUESTION , CAN IT BE POSSIBLE IF YES,
KINDLY PROVIDE SOME SOLUTION TO GO AHEAD.

**Reply by JesusPoderoso** [reply]
Hi @Akul1903 

> And One more doubt By any chance fastdds is a license product or a free open source product

Fast DDS uses the [Apache License](https://www.apache.org/licenses/LICENSE-2.0). Refer to the [license file](https://github.com/eProsima/Fast-DDS/blob/master/LICENSE) for more information.

> I am working on one publisher and 2 subscriber model in which each subscriber has different data members.

Publisher and subscriber need the same data type to match each other. If they have different data types, they will not match.
As a solution, the subscribers can receive all data and process only the desired parameters. The [content filtered topic](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/contentFilteredTopic/contentFilteredTopic.html) may be of interest to you.

**Reply by Akul1903** [reply]
hi @JesusPoderoso Thanks for confirming 

I have one more doubt:

Can subscriber send data back to publisher?
if is it possible then how?

**Reply by JesusPoderoso** [reply]
Hi @Akul1903.
No, you would need to create a new set of publisher-subscribers. 
This [request-reply](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/request_reply/request_reply.html) documentation section may interest you.
